### ISSUE TITLE
Release 3.1.1

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#1546]
I'm afraid we will need to issue a Release 3.1.1.

I found a bug with Giftcard.
When you create a new one and save it, there is an error thrown and the dialog box doesn't close.

**Comment by daN4cat** [comment]
I need to check if there are more similar issues, somehow something went wrong in recent time.
Time to retest everything before seal a 3.1.1.

**Comment by daN4cat** [comment]
Also found out that some en_GB translations for Giftcard are not correct, fixing those too in file and on Weblate to keep them in sync.

**Comment by odiea** [comment]
Which Giftcard setting are you using?  It looks ok using generate in series.

**Comment by daN4cat** [comment]
@jekkos there is something strange with Weblate, it messed up quite some strings and replaces all the "You have successfully deleted {Gifcard, Employee, Customer, etc.}" with the same string all over the different modules. Basically I've got all the came sentence with Giftcard. I'm changing in GitHub but you need to sync back to Weblate and check why this link between strings.

**Comment by daN4cat** [comment]
Create a new card with random generation. I got an error and there is a bug in the code.

**Comment by odiea** [comment]
And you have Gift Card number set to generate random. I do not get this issue.

**Comment by daN4cat** [comment]
Also I just found out that tax calculation with tax included (VAT 20%) doesn't appear to be right in reports, I mean the total at least. Oh well.... this tax story, how much I love it!

**Comment by daN4cat** [comment]
I'm puzzled by the VAT mistake I saw. It might be an issue moving the old database to the new version. So something wrong with my migration?

**Comment by objecttothis** [comment]
Not sure if this is new to 3.1.0 but when changing the currency symbol manually it throws a HTTP 400 error after hitting tab "http://domain/subfolder/public/[object Object]" The type of request is post.  I have tested this on both my server and an external server to eliminate my server as being the culprit.

**Comment by objecttothis** [comment]
I'm also noticing on two different servers that the .htaccess mod_rewrite is not working to forward to the /public/ folder.  In both cases I am using a subdirectory.  I will look into why it's not working and submit a pull request.

**Comment by daN4cat** [comment]
The migrate function is not working correctly with Tax included (VAT) case.

**Comment by daN4cat** [comment]
@objecttothis I've got a problem with an Apache option to control syblink race condition exploits.
Make sure you don't any such thing enabled when testing. I disabled even because that Apache patch is not recommended in any case.

**Comment by SteveIreland** [comment]
@daN4cat Did you want me to take a look at the migration issue this week?  It will take me some time because I'm already booked to my eyebrows with commitments this weekend plus I'll need to set  up 3.0.2 for a VAT scenario and create some test data to work with.

**Comment by objecttothis** [comment]
localization is also messed up for az and az_AZ. For az it populates the currency symbol as ¤ but it should be ₼ (https://en.wikipedia.org/wiki/Azerbaijani_manat) and the currency side should be right, not left.  for az_AZ and az-AZ it makes the currency symbol "man." which is the old currency.  the currency code is actually AZN.  In either case the symbol should be ₼.  I believe this was correct in earlier master's but somehow got broken.  Are we pulling this information from an external source or do we have that information stored within the app?

**Comment by objecttothis** [comment]
@daN4cat one test environment I am using is running nginx and the other apache.  On the apache build I will look into this option to disable it.

**Comment by daN4cat** [comment]
@SteveIreland can you please help? I'm getting lost in the nesting of code for this Tax calculation story.
The migration routine doesn't calculate the correct tax in tax included case.

Example: Tax is 20% (VAT).
Total 50
Calculate tax is 10, where it should be 8.33

**Comment by SteveIreland** [comment]
Will do.

**Comment by objecttothis** [comment]
@daN4cat I checked on my apache install that mod_ruid was not enabled and I changed the Options directive from +SymlinkIfOwnersMatch to +FollowSymLinks and then restarted apache.  That did not have any effect.

**Comment by daN4cat** [comment]
@objecttothis make sure you use PHP 7.1 so libraries are latest. The currency support is a php extension story.

**Comment by objecttothis** [comment]
on one server I don't have multiple php installs setup and need 5.6 for other php apps run on the server.  on the other server I changed the php to 7.0 (7.1 wasn't available) but it didn't make any difference.  both have the intl extension installed and updated.  Is it a different extension or is that the one that the information is being drawn from?

**Comment by daN4cat** [comment]
@objecttothis if you go here: http://php.net/manual/en/numberformatter.formatcurrency.php

Somebody from Colombia is saying that php_intl is not totally correct but he gives a solution:

```
For example "COP" (Colombian peso) if defined to use 2 decimals, but NumberFormat::formatCurrency() use 0 decimal for this currency (I do not know why!).

Here is the code I use :

$fmt = new \NumberFormatter( 'fr', \NumberFormatter::CURRENCY);
$fmt->setTextAttribute( $fmt::CURRENCY_CODE, 'COP' );
$fmt->setAttribute( $fmt::FRACTION_DIGITS, 2 );
$numberString = $fmt->format( 1234.56 );

The output is: 1 234,56 $CO

If locale change to 'en' then the output is : COP1,234.56
```

**Comment by daN4cat** [comment]
@SteveIreland I know you are busy with the backoffice feature, but please find the time to fix the migration issue I mentioned above. Otherwise we get stuck to 3.1.1.
Also I hope we can add corrections to en-US translations if any. And any other small fix.

**Comment by objecttothis** [comment]
I noticed that en-US doesn't have changes I made in weblate. have those just not been pulled down yet (e.g. localisation should be localization)

**Comment by objecttothis** [comment]
it appears PHP intl relies on ICU for it's localization. I'll look into making sure that the most recent version is installed and correcting inaccuracies in their database.

**Comment by daN4cat** [comment]
Good.

WRT en-US please check Weblate. I noticed strange changes of translations by the tool. I ended up changing translations I Github to get what I wanted. So maybe you can do that too and then @jekkos syncs Weblate with what it is in Github.

**Comment by SteveIreland** [comment]
I noticed that there is now a subfolder of public in the general release named bower_components.  I don't believe that was there in the earlier official releases.   I've removed it and all still seems  okay.  Should it be there?

**Comment by SteveIreland** [comment]
I just finished running the upgrade script from a really basic 3.0.2 database and received the following SQL error.  I've encountered it a few times but dismissed it because I just thought I ran something out of sequence (a common occurrence for me).  But since I'm upgrading an unadulterated data set and still got the message I thought I should raise a flag.

```
SQL query:


INSERT INTO `ospos_app_config` (`key`, `value`) VALUES
('barcode_formats', '[]'),
('barcode_font', 'Arial'),
('barcode_font_size', '10'),
('barcode_height', '50'),
('barcode_quality', '100'),
('barcode_type', 'Code39'),
('barcode_width', '250'),
('receiving_calculate_average_price', '0')
MySQL said: Documentation

#1062 - Duplicate entry 'barcode_font' for key 'PRIMARY'
```

**Comment by SteveIreland** [comment]
Same for
```
SQL query:


ALTER TABLE ospos_giftcards ADD record_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP
MySQL said: Documentation

#1060 - Duplicate column name 'record_time'
```

**Comment by daN4cat** [comment]
Oh good. I had the same running my upgrade from an old data setup and thought something was wrong in my other scripts.

I'll remove those lines as that stuff was added recently by @objecttothis and thought we missed something.

**Comment by daN4cat** [comment]
Are those the only errors?

**Comment by daN4cat** [comment]
Bower_components it's just for development not production.

@jekkos something to fix in the script.

**Comment by jekkos** [comment]
@objecttothis Currency symbol change might be a jQuery migration issue..

**Comment by jekkos** [comment]
quite a couple of things to fix here I see..

**Comment by objecttothis** [comment]
@daN4cat the reason why we are getting those errors on adding those lines in the 3.0.2_to_3.1.0 upgrade script is that those changes were made in previous upgrades.  We (I) need to figure out where those features (barcode, etc) were introduced and put them in those upgrade scripts rather than the 3.0.2_3.1.0.  Does that make sense?

**Comment by daN4cat** [comment]
Tax migration issue fixed and in master and I removed the clashing db upgrade script issues.
I renamed to be 3.0.2 to 3.1.1. to avoid issue with the other one.

@objecttothis to be honest I'm not sure we missed something as other people upgraded to 3.0.2 with no issue. Also the upgrade from 3.0.2 to 3.1.0 was ok and we could see that.

Are you sure you didn't miss some other step in the upgrade? I need to ask twice questions here to everybody so we make sure we don't introduce more regressions.

**Comment by daN4cat** [comment]
@jekkos with jQuery would influence the currency symbol? Not sureI get that...

**Comment by objecttothis** [comment]
@daN4cat I haven't looked at the code but if something was changed and it's not properly handling Unicode characters that would explain why the manat symbol wasn't displaying correctly where it was before in the az locale

**Comment by objecttothis** [comment]
@daN4cat I can pull another master from before my changes to the upgrade script and run through the upgrade process again from 2.0.2, but a simple grep search will show whether those keys are added in any other upgrade scripts. another solution would be to change the SQL statement to use the ignore keyword which will just skip adding the key for people who already have it.

**Comment by daN4cat** [comment]
Yes, a grep search could do indeed.

**Comment by daN4cat** [comment]
Barcode stuff was introduced with 2.3.1_to_2.3.2.sql

**Comment by objecttothis** [comment]
thanks. our son is home now. I can work on it in the morning but we can always bump it to 3.2.0 if we need to get 3.1.1 before

**Comment by daN4cat** [comment]
fixing....

(glad to know your son is home. Surely he will have a better recovery there then in an hospital)

**Comment by daN4cat** [comment]
``record_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP, ` was introduced with PR #11

**Comment by daN4cat** [comment]
Found and now fixed the barcode and giftcard recordtime in the right upgrade scripts.

**Comment by saruban** [comment]
Finally finished? 
Waiting for 3.1.1

**Comment by SteveIreland** [comment]
Can one ever say that a work of organic art is truly finished?  I think not.

**Comment by daN4cat** [comment]
somehow organic sounds to me like implying worms, hope this is not a "can of worms" then :-)

**Comment by daN4cat** [comment]
@objecttothis I know it's a big ask given your personal circumstances, but do you think you can work on translating en_US straight on GitHub and give a try to the upgrade process?
It should be all fine (finger crossed) and hopefully we can release 3.1.1 by end of the week.

I would need to check if the traybin zip contains development stuff or not.

**Comment by jekkos** [comment]
Changed some of the older jQuery POSTs to work with 2.0. Think mailChimp config was broken as well

**Comment by jekkos** [comment]
For my understanding, what is the source of truth now? I see there are still translations pending in weblate. Idea is to remove translations there and add them again using the ospos locale hyphen..

**Comment by jekkos** [comment]
{code};
++<<<<<<< 887c8374b650743751f2884d26a6a84f2eb9c367
 +$lang["customers_none_selected"] = "You have not selected any Customer(s) to delete";
 +$lang["customers_one_or_multiple"] = "Customer(s)";
 +$lang["customers_quantity"] = "Quantity";
 +$lang["customers_stats_info"] = "Stats";
 +$lang["customers_successful_adding"] = "You have successfully added Customer";
 +$lang["customers_successful_deleted"] = "You have successfully deleted Customer";
++=======
+ $lang["customers_none_selected"] = "You have not selected any customer(s) to delete";
+ $lang["customers_one_or_multiple"] = "Customer(s)";
+ $lang["customers_quantity"] = "Quantity";
+ $lang["customers_stats_info"] = "Stats";
+ $lang["customers_successful_adding"] = "You have successfully added a customer";
+ $lang["customers_successful_deleted"] = "You have successfully deleted a Giftcard";
++>>>>>>> Translated using Weblate (English (United Kingdom))

{code}

I'm about to resolve a  merge conflict while rebasing weblate with latest master. @daN4cat which version is the correct one? With or without caps? I saw you changed some sentences in weblate

**Comment by daN4cat** [comment]
@jekkos the version in Github is the correct one. Weblate is wrong.

**Comment by objecttothis** [comment]
When I go to weblate and click on the en_US Sales translation I get


Request Method: | GET
-- | --
http://weblate.jpeelaer.net/projects/ospos/sales/en_US/
1.9
RepositoryException
fatal: HEAD does not point to a branch (128)
/usr/local/lib/python2.7/dist-packages/weblate/trans/vcs.py in _popen, line 174
/usr/bin/python
2.7.10
['/usr/local/bin',  '/usr/lib/python2.7',  '/usr/lib/python2.7/plat-x86_64-linux-gnu',  '/usr/lib/python2.7/lib-tk',  '/usr/lib/python2.7/lib-old',  '/usr/lib/python2.7/lib-dynload',  '/usr/local/lib/python2.7/dist-packages',  '/usr/lib/python2.7/dist-packages']
Tue, 5 Sep 2017 20:41:15 +0200

**Comment by objecttothis** [comment]
@daN4cat I am getting ready to go to bed, but I should be able to work on it in the morning (GMT+4) and I'd imagine that I can have it done by the end of the week or sooner.  To make sure I understand what we are asking, I need some clarification:
- I completed the translation of en_US on weblate, but left a few phrases marked "needs review" to make sure there were no objections. did the translation get pulled into the master from weblate, or do I need to do it again?
- I think you are asking me to create a patch and submit a pull request on Github and just bypass weblate.  Can we make sure that weblate isn't going to overwrite my changes?
- Yes, I will download the latest master and give the upgrade process a try.  I wonder if there is a way that we can number the upgrade files to make it extremely clear what order they are supposed to be performed in.  Generally it's clear, but for example, it wasn't intuitive whether the 2.3_migrate_locations.sql should be performed before or after 2.2.2_to_2.3.sql... I wonder if we shouldn't just merge 2.3_migrate_locations into 2.2.2_to_2.3.sql.

**Comment by daN4cat** [comment]
Write the correct translations in Github language/en-US, @jekkos will sync Weblate from that. Good night.

**Comment by jekkos** [comment]
You're getting the error because weblate expects the files to be in en_US and not en-US. Think I will move the ospos version first to en_US, then apply your changes and then change all to en-US in both weblate and git

**Comment by daN4cat** [comment]
I removed en_US from master, so that should not be available anymore. Only en-GB and en-US

**Comment by objecttothis** [comment]
OK, then today I am working on correcting the en-US translation in Github.

**Comment by jekkos** [comment]
@objecttothis  so I can completely discard all pending changes in weblate then? Otherwise you'll  have to wait for me to sync once again and you can proceed?

**Comment by objecttothis** [comment]
I already did the translation in weblate so if you could transfer everything from the weblate translation to en-US in GitHub that would save me a lot of time.

**Comment by objecttothis** [comment]
OK, I just pulled the current master and installed two instances side-by-side.  One has a clean database using the database.sql and the other has my database upgraded by running each of the scripts in order.  I am happy to say that the two database structures are identical and the key names in ospos_app_config are identical.  Someone needs to test 3.0.2_to_3.1.1.sql in the current master against their 3.0.2 install.  If there are errors they are likely because we need to put those lines in the proper update script, so report them so I can submit a pull request.

**Comment by objecttothis** [comment]
Azerbaijani does not appear in the drop-down for language under localization in App Config

**Comment by objecttothis** [comment]
The localization currency symbol for az is currently still broken.
![az](https://user-images.githubusercontent.com/17935339/30109617-0dc6b140-9318-11e7-8b8f-ccc51a32c1bb.JPG)
However az-AZ and az_AZ are the correct symbol now. 
![az-az](https://user-images.githubusercontent.com/17935339/30109616-0dc5c078-9318-11e7-80be-84cbc1df0591.JPG)

@jekkos thanks for whatever you did.  Any idea why the symbol isn't appearing properly with az? Also, any idea why the symbol is appearing on the left?  It's supposed to be on the right side for all three locales.  Is ospos dictating that or is ospos pulling that information from PHP intl?

**Comment by jekkos** [comment]
could be an encoding issue along the way, not sure what encoding the http reply for this token is using exaclty.. also the doctype specifies utf-8 I suppose so don't know if this token is included

information on left or right is taken from php intl, from which I assumed things would be correct.. maybe search the net for similar bug reports there.. there used to be an option to set the symbol on either side but I decided to  remove that in favor of the intl currency locale.

**Comment by jekkos** [comment]
Regarding weblate the translations there went in merge conflict which I need to resolve without some of my beloved textmode tools in a docker container commandline

**Comment by objecttothis** [comment]
@jekkos, then do you want me to wait until you do that and pull down to github, or do you want me to proceed with working on the translation of en_US on github.  I'm ok if you do, but I wanted to wait to hear from you rather than do the work and have it overwritten.

**Comment by jekkos** [comment]
The weblate version of en_US is back in en_US next to en-US (which contains the github version)..

**Comment by jekkos** [comment]
So but then I guess en_US will contain latest and can overwrite en-US? in that case I need to adapt the mapping in weblate to refer to translations in en-US folder as those are the ones currently used

**Comment by objecttothis** [comment]
This is a little confusing for me.  My understanding is that we settled on en-US being the format that we would use and that en-US would be the source language.  Is there a way in weblate to change en_US to en-US and make that the source, then once that is done bring en-US down to github en-US?

**Comment by objecttothis** [comment]
in the interest of getting 3.1.1 out, my vote is to punt on the currency symbol showing on the correct side in the az-AZ, az_AZ and az locales.  That way we can focus on getting the bigger issues resolved.

**Comment by daN4cat** [comment]
Now GitHub en_US needs to override en-US and then remap Weblate to en-US

**Comment by daN4cat** [comment]
And delete GitHub en_US (yet again)

**Comment by daN4cat** [comment]
@jekkos would you do that or shall I do the shift in Github this eve and then you fix Weblate?

**Comment by objecttothis** [comment]
ok, let me know when it's done and I will check the translation of en-US on weblate so we can get this bad boy out the door.

**Comment by jekkos** [comment]
Ok do it in github already. I'll pull those changes back into weblate and then adapt the config to use en-US

**Comment by daN4cat** [comment]
Guys, there is no difference between en_US and en-US translations. Something is weird with Weblate or the correct translations are somewhere else.
@jekkos any clue?

**Comment by daN4cat** [comment]
@jekkos I submitted few more fixes to en-GB in Github. Sorry for the pain but I find the inverted process quicker as the Weblate to Github propagation time is a bit too long at the moment.

**Comment by objecttothis** [comment]
@daN4cat @jekkos ok, I will finish the translation of en_US today in github and submit the pull request

**Comment by objecttothis** [comment]
I just finished and submitted the pull request for the en-US on github.  When merging and then subsequently fixing weblate please be careful not to overwrite the github translation with weblate.

**Comment by objecttothis** [comment]
What else is needed to release 3.1.1?

**Comment by daN4cat** [comment]
I merged the translations and amended en-GB accordingly since I noticed that there were few sentences that didn't make full sense (spotted thanks to @objecttothis changes).

**Comment by daN4cat** [comment]
With regards to release 3.1.1, I just need to wait for the csv issue to be cleared.

**Comment by daN4cat** [comment]
If no further, then I can release 3.1.1 on Saturday.

**Comment by tenkodoy** [comment]
What about addition of expiry dates and expiry notifications?

**Comment by daN4cat** [comment]
@tenkoboy if somebody wants to contribute it it will be for 3.2.0

**Comment by jekkos** [comment]
Weblate seems quite dead atm, after removing source language folder it surely needs some love..

**Comment by sanubari97** [comment]
how can i correct/contribute translation for my language ?

**Comment by objecttothis** [comment]
once @jekkos gets weblate back up and running you will be able to sign up for an account there and contribute.

**Comment by daN4cat** [comment]
@sanubari97 I'm going to release 3.1.1. Your translations will be part of 3.2.0.
However when they are move from Weblate to GitHub you can just take those files and manually upgrade 3.1.1, it will not break anything because 3.2.0 just adds more translations.

**Comment by daN4cat** [comment]
Once the build in bintray is ready I will add it and release 3.1.1

**Comment by daN4cat** [comment]
OK, 3.1.1 is now released.

**Comment by sanubari97** [comment]
how to upgrade from 3.1.1?

**Comment by daN4cat** [comment]
There is no need for db upgrade from 3.1.0 to 3.1.1.
Just upgrade the code using the long name zip file.
